### PR TITLE
Fix and improve unit5 REINFORCE computation of the returns

### PR DESCRIPTION
**Pull request by Chris1nexus**
### Theory
The provided implementation computes the policy loss considering only the return G_0, as:
sum over all t of the G_0 *log_policy(a_t|s_t)

However, the reinforce algorithm requires to compute the returns G_t for all t,
to then calculate the sum over all t of the G_t *log_policy(a_t|s_t).

Hence, the wrong computation has been fixed to consider the discounted returns at each timestep.

### Implementation improvements upon pytorch
The provided implementation improves also upon pytorch official REINFORCE implementation by
using a python deque to store returns, which allows for O(1) list insertion at the front of the list of the returns.
This is an impactful computation overhead for long episodes.
Their original implementation employs a standard python list, which has O(N) complexity for insertions at beginning of a list,
(N is its size).

Time complexity reference: https://docs.python.org/3/tutorial/datastructures.html#using-lists-as-queues 
Pytorch Reinforce: https://github.com/pytorch/examples/blob/main/reinforcement_learning/reinforce.py#L68

### Comments
Note: the policy loss has still not been multiplied by the constant 1/n_step, to maintain the least number of changes to obtain the correctly working implementation (a scaling constant would be accounted for by the learning rate anyway in the optimization process)

Further Note: this merge request is a follow-up to https://github.com/huggingface/deep-rl-class/pull/70#issue-1315886856

### CODE REVIEW

**Comment by simoninithomas** [comment]
Hey there 👋 . Thanks for pointing this out 🤗 . The Reinforce algorithm unit needs to be updated. I'm checking this implementation and your comment this week. Keep you updated 🤗

**Comment by Chris1nexus** [comment]
Ok thanks for taking the time to review it!

**Comment by simoninithomas** [comment]
Hey there 👋  I'm merging it since we launch Policy Gradient for the v2.0 version using your update. Thanks for your help 🤗

**Comment by simoninithomas** [comment]
(I mention you in the new notebook that will be published tomorrow 🤗 ) thanks again for the update

**Comment by Chris1nexus** [comment]
glad to be of help for this one! 
thanks also for the mention @simoninithomas !
